src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -596,7 +596,7 @@ private function createUserProviders($config, ContainerBuilder $container)
         $providerIds = array();
         foreach ($config['providers'] as $name => $provider) {
             $id = $this->createUserDaoProvider($name, $provider, $container);
-            $providerIds[] = $id;
+            $providerIds[str_replace('-', '_', $name)] = $id;
```

Comment:
by now i think the conversion should happen in `getUserProviderId` actually :thinking: and simply make the check below `$userProvider`. That would be more robust.

```php
$userProvider = $this->getUserProviderId($firewall[$key]['provider']);
if (!isset($providerIds[$userProvider])) {
   // throw
}
```

sorry :)

Reply:
not sure about BC :)

Reply:
point is we create a different id 2 lines above by passing `$name` as is to `createUserDaoProvider`.. that looks weird.

Reply:
I get your point but yes, changing this is not fine regarding BC. Note also that when using a custom provider (specifying `my_provider: { id: some_service }`), `createUserDaoProvider` returns the `some_service` id, it doesn't rely on `getUserProviderId`. Let's avoid any behavior change here, not the right moment for that :) 
I'm going to review this part of the extension in 2.7 once this merged, I believe there are some bugs. Right now, 3.4's security setup is broken, this covers all edge cases and makes it work again. I would like to quickly move on and polish afterwards if needed.
Btw, thanks for your review, it was useful actually.